Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -37,60 +42,242 @@ open scoped Matrix
+  rw [hv', dotProduct_smul, IsROrC.real_smul_eq_coe_mul, ← mul_zero ↑t] at aux
+  exact hv <| Matrix.dotProduct_star_self_eq_zero.mp <| mul_left_cancel₀
+    (IsROrC.ofReal_ne_zero.mpr ht) aux
+
+lemma sqrt_sq : (hA.pow 2 : PosSemidef (A ^ 2)).sqrt = A :=
+  (hA.pow 2).posSemidef_sqrt.eq_of_sq_eq_sq hA (hA.pow 2).sq_sqrt
+
+lemma eq_sqrt_of_sq_eq {B : Matrix n n 𝕜} (hB : PosSemidef B) (hAB : A ^ 2 = B) : A = hB.sqrt := by
+  subst B
+  rw [hA.sqrt_sq]
+
+end sqrt
+
+end PosSemidef
+
 @[simp]
 theorem posSemidef_submatrix_equiv {M : Matrix n n R} (e : m ≃ n) :
     (M.submatrix e e).PosSemidef ↔ M.PosSemidef :=
   ⟨fun h => by simpa using h.submatrix e.symm, fun h => h.submatrix _⟩
 #align matrix.pos_semidef_submatrix_equiv Matrix.posSemidef_submatrix_equiv
 
 /-- The conjugate transpose of a matrix mulitplied by the matrix is positive semidefinite -/
-theorem posSemidef_conjTranspose_mul_self (A : Matrix m n R) : Matrix.PosSemidef (Aᴴ * A) := by
+theorem posSemidef_conjTranspose_mul_self (A : Matrix m n R) : PosSemidef (Aᴴ * A) := by
   refine ⟨isHermitian_transpose_mul_self _, fun x => ?_⟩
   rw [← mulVec_mulVec, dotProduct_mulVec, vecMul_conjTranspose, star_star]
   exact Finset.sum_nonneg fun i _ => star_mul_self_nonneg _
 
 /-- A matrix multiplied by its conjugate transpose is positive semidefinite -/
-theorem posSemidef_self_mul_conjTranspose (A : Matrix m n R) : Matrix.PosSemidef (A * Aᴴ) :=
+theorem posSemidef_self_mul_conjTranspose (A : Matrix m n R) : PosSemidef (A * Aᴴ) :=
   by simpa only [conjTranspose_conjTranspose] using posSemidef_conjTranspose_mul_self Aᴴ
 
-/-- The eigenvalues of a positive semi-definite matrix are non-negative -/
-lemma PosSemidef.eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
-    (hA : Matrix.PosSemidef A) (i : n) : 0 ≤ hA.1.eigenvalues i :=
-  (hA.re_dotProduct_nonneg _).trans_eq (hA.1.eigenvalues_eq _).symm
-
 lemma eigenvalues_conjTranspose_mul_self_nonneg (A : Matrix m n 𝕜) [DecidableEq n] (i : n) :
     0 ≤ (isHermitian_transpose_mul_self A).eigenvalues i :=
-  (Matrix.posSemidef_conjTranspose_mul_self _).eigenvalues_nonneg _
+  (posSemidef_conjTranspose_mul_self _).eigenvalues_nonneg _
 
 lemma eigenvalues_self_mul_conjTranspose_nonneg (A : Matrix m n 𝕜) [DecidableEq m] (i : m) :
     0 ≤ (isHermitian_mul_conjTranspose_self A).eigenvalues i :=
-  (Matrix.posSemidef_self_mul_conjTranspose _).eigenvalues_nonneg _
+  (posSemidef_self_mul_conjTranspose _).eigenvalues_nonneg _
+
+/-- A matrix is positive semidefinite if and only if it has the form `Bᴴ * B` for some `B`. -/
+lemma posSemidef_iff_eq_transpose_mul_self [DecidableEq n] {A : Matrix n n 𝕜} :
+    PosSemidef A ↔ ∃ (B : Matrix n n 𝕜), A = Bᴴ * B := by
+  refine ⟨fun hA ↦ ⟨hA.sqrt, ?_⟩, fun ⟨B, hB⟩ ↦ (hB ▸ posSemidef_conjTranspose_mul_self B)⟩
+  simp_rw [← PosSemidef.sq_sqrt hA, pow_two]
+  rw [hA.posSemidef_sqrt.1]
+
+lemma posSemidef_of_eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
```

Comment:
```suggestion
lemma PosSemidef.of_eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
```
will give `(.of_eigenvalues_nonneg _ _)` dot notation, instead of no dot notation.

Reply:
oops, I meant to call it `Matrix.IsHermitian.posSemidef_of_eigenvalues_nonneg`. Sorry!